File: typing_trainer.go
// typing_trainer.go — Go версия

package main

import (
	"bufio"
	"fmt"
	"math/rand"
	"os"
	"strconv"
	"strings"
	"time"
)

func generateSequence(length int, usePunctuation bool) string {
	chars := "0123456789"
	if usePunctuation {
		chars += "!@#$%^&*()-_=+[]{};:,.<>/?`~"
	}
	b := make([]byte, length)
	for i := range b {
		b[i] = chars[rand.Intn(len(chars))]
	}
	return string(b)
}

func printHighlight(target, userInput string) {
	fmt.Print("\r\033[K") // очистка строки
	for i := 0; i < len(target); i++ {
		var ch byte
		var color string
		if i < len(userInput) {
			ch = userInput[i]
			if ch == target[i] {
				color = "\033[32m"
			} else {
				color = "\033[31m"
			}
		} else {
			ch = target[i]
			color = "\033[33m"
		}
		fmt.Printf("%s%c\033[0m ", color, ch)
	}
	fmt.Print("\n")
}

func main() {
	rand.Seed(time.Now().UnixNano())

	fmt.Println("\033[36m⌨️  Тренажёр печати (цифры) (Go)\033[0m")
	fmt.Print("Длина упражнения (по умолч. 20): ")
	reader := bufio.NewReader(os.Stdin)
	lengthStr, _ := reader.ReadString('\n')
	lengthStr = strings.TrimSpace(lengthStr)
	length := 20
	if lengthStr != "" {
		if l, err := strconv.Atoi(lengthStr); err == nil && l > 0 {
			length = l
		}
	}

	fmt.Println("Нажмите Enter, когда будете готовы...")
	reader.ReadString('\n')

	target := generateSequence(length, false)
	fmt.Printf("\033[33mВведите:\033[0m %s\n", strings.Join(strings.Split(target, ""), " "))

	fmt.Println("Вводите цифры (по одной, без Enter). Для выхода нажмите Ctrl+C.")
	userInput := ""

	// Для чтения посимвольно без Enter используем пакет "github.com/eiannone/keyboard"
	// Но для упрощения используем построчный ввод
	// Попробуем читать посимвольно с помощью bufio с буфером, но в Go сложно без внешних библиотек.
	// В этой версии используем построчный ввод (Enter).
	fmt.Println("Вводите цифры последовательно (Enter после каждой):")
	start := time.Now()
	for len(userInput) < len(target) {
		fmt.Print("> ")
		ch, _ := reader.ReadString('\n')
		ch = strings.TrimSpace(ch)
		if len(ch) == 0 {
			continue
		}
		userInput += ch[:1]
		printHighlight(target, userInput)
	}
	elapsed := time.Since(start).Seconds()

	correct := 0
	for i := 0; i < len(target) && i < len(userInput); i++ {
		if userInput[i] == target[i] {
			correct++
		}
	}
	accuracy := float64(correct) / float64(len(target)) * 100
	wpm := float64(len(target)) / (elapsed / 60)

	fmt.Printf("\n\033[36mСтатистика:\033[0m\n")
	fmt.Printf("  Время: %.1f сек\n", elapsed)
	fmt.Printf("  Скорость: %.1f зн/мин\n", wpm)
	fmt.Printf("  Точность: %.1f%%\n", accuracy)
	fmt.Printf("  Ошибок: %d\n", len(target)-correct)

	// Сохранение
	f, _ := os.OpenFile("typing_stats.txt", os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
	defer f.Close()
	fmt.Fprintf(f, "Go\t%.1f\t%.1f\t%.1f\t%d\n", elapsed, wpm, accuracy, len(target)-correct)
}

File: typing_trainer.py


### 1. `typing_trainer.py` (Python)

```python
# typing_trainer.py — Python версия

import random
import time
import sys
from colorama import init, Fore, Style
try:
    import keyboard  # для чтения клавиш без Enter
except ImportError:
    keyboard = None

init(autoreset=True)

def generate_sequence(length=20, use_punctuation=False):
    chars = '0123456789'
    if use_punctuation:
        chars += '!@#$%^&*()-_=+[]{};:,.<>/?'
    return ''.join(random.choice(chars) for _ in range(length))

def print_highlight(target, user_input):
    """Печатает строку, выделяя правильные/неправильные символы."""
    result = []
    for i, (t, u) in enumerate(zip(target, user_input)):
        if u == t:
            result.append(Fore.GREEN + u + Style.RESET_ALL)
        else:
            result.append(Fore.RED + u + Style.RESET_ALL)
    # Дописываем остаток target, если ввод короче
    if len(user_input) < len(target):
        for ch in target[len(user_input):]:
            result.append(Fore.YELLOW + ch + Style.RESET_ALL)
    print(' '.join(result))

def main():
    print(f"{Fore.CYAN}⌨️  Тренажёр печати (цифры) (Python)")
    length = input("Длина упражнения (по умолч. 20): ").strip()
    length = int(length) if length else 20

    print("Нажмите Enter, когда будете готовы...")
    input()

    target = generate_sequence(length)
    print(f"{Fore.YELLOW}Введите:{Style.RESET_ALL} " + ' '.join(target))

    start = time.time()
    user_input = []

    # Используем keyboard для посимвольного ввода (если доступно)
    if keyboard:
        print("Вводите цифры (нажмите Esc для выхода):")
        while len(user_input) < len(target):
            event = keyboard.read_event(suppress=True)
            if event.event_type == 'down':
                if event.name == 'esc':
                    print("\nПрервано.")
                    return
                if event.name.isdigit():
                    user_input.append(event.name)
                    sys.stdout.write('\r' + ' ' * 80 + '\r')
                    print_highlight(target, ''.join(user_input))
                    sys.stdout.flush()
                elif event.name == 'backspace' and user_input:
                    user_input.pop()
                    sys.stdout.write('\r' + ' ' * 80 + '\r')
                    print_highlight(target, ''.join(user_input))
                    sys.stdout.flush()
    else:
        # fallback: построчный ввод
        print("Вводите цифры последовательно (Enter после каждой):")
        while len(user_input) < len(target):
            ch = input("> ").strip()
            if not ch:
                continue
            user_input.append(ch[0])
            sys.stdout.write('\r' + ' ' * 80 + '\r')
            print_highlight(target, ''.join(user_input))
            sys.stdout.flush()

    elapsed = time.time() - start
    user_str = ''.join(user_input)

    # Статистика
    correct = sum(1 for t, u in zip(target, user_str) if t == u)
    total = len(target)
    accuracy = (correct / total) * 100 if total > 0 else 0
    wpm = (total / (elapsed / 60)) if elapsed > 0 else 0

    print(f"\n{Fore.CYAN}Статистика:")
    print(f"  Время: {elapsed:.1f} сек")
    print(f"  Скорость: {wpm:.1f} зн/мин")
    print(f"  Точность: {accuracy:.1f}%")
    print(f"  Ошибок: {total - correct}")

    # Сохраняем результат в файл
    with open("typing_stats.txt", "a", encoding="utf-8") as f:
        f.write(f"Python\t{elapsed:.1f}\t{wpm:.1f}\t{accuracy:.1f}\t{total - correct}\n")

if __name__ == "__main__":
    main()
